### ISSUE TITLE
Fix citation detection for arXiv paper `1508.07252`

### ISSUE THREAD

**Issue by andrewhead** [allenai/scholar-reader#54]
In our changes that enables the detection of separate citations separately (#24), we employed a new method of citation detection that does not apply to a paper we often use in examples---[arXiv paper 1508.07252](https://arxiv.org/abs/1508.07252). I think the reason is that it uses different internal macros for generating citation text than the standard LaTeX ones our new colorization method detects.

**Comment by andrewhead** [comment]
Fro paper `1911.00568`, see this section in the `cite` package macros (hosted here http://tug.ctan.org/tex-archive/macros/latex/contrib/cite/cite.sty), starting here:

```latex
% For each citation, check if it is defined. If so, then extract plain
% value to \@B@citeB (without hyperlink info). Then, 
```

**Comment by andrewhead** [comment]
The trick in supporting documents formatted with the `cite` package is that citations get checked to see if they are a number (`posnumtest`), and if they are, they're added to a sorted list of numbers. Then this sorted list of numbers is printed.

On the one hand, this makes me think it might be complicated to figure out where to insert our color macros. On the other hand, hyperref finds a way to insert its links into these sorted lists (see again paper 1911.00568), so we should be able to do the same.

**Comment by andrewhead** [comment]
In general, we should be able to rely on many papers with our existing macros because hyperref is injected into many by AutoTeX.

From the log for paper 1911.00568, it appears that `\@citecolor` is getting called, as our macros expect. See this excerpt of the log:

```
\Hy@colorlink #1->\begingroup 
#1<-\@citecolor 
```

The question is, why isn't `\@citecolor` getting resolved using the macro we defined? Does it get overwritten somewhere else in the document? That's what I need to look for next.

**Comment by andrewhead** [comment]
This ended up being a much more involved fix than I anticipated. As noted above, sometimes `\@citecolor` was not called for coloring the text.

The fixed turned out to be to instead instrument the more general purpose hyperref internal command called `hyper@link@` to check if it was processing an citation and, if so, figure out the citation key and try to color the text that a link is getting applied to.

This required shims that were different for different types of papers. For `1508.07252`, the macros needed to be imported after `\begin{document}`, the main reason being that AutoTeX sometimes adds a `\RequirePackage{hyperref}` command right before `\begin{document}` that clobbered our redefinition of `hyper@link@`. For `1911.00568` (which used package `cite`, I think), `\@citeb`  couldn't be used to check which key was getting output, so we then had to come up with another method for trying to guess the citation key based on the parameters to `hyper@link@`.

For the time being, colorization works on the test cases we've tested on, though we may need to reopen this in the future to provide even more robust citation coloring commands.

See the fixes to this issue in https://github.com/allenai/scholar-reader/commit/3ede8c2d41474ac7f25fab0bc7d5e6ecf6c7b447

**Comment by andrewhead** [comment]
Also, I no longer need these commands for coloring, which were used in an initial version of my fixes for paper `1508.07252`. However, there's a chance these will be useful in the future, and perhaps even maybe the starting point for a more generalizable citation coloring solution. Here's the snippet. This would be placed inside the `scholarregistercolor` macro.

```latex
% * Support dynamic coloring in IEEE LaTeX formats:
% For IEEE LaTeX formats, override the b@<citation-key> macros that are 
% used to look up the text for a citation. Override those macros to color the 
% citation text. This is the part of the 'scholarregistercolor' macro that
% requires the whole macro to be run after '\begin{document}'. If included 
% before '\begin{document}, citation coloring doesn't work, I assume because
% the 'b@' macros aren't defined until the beginning of the document.
% While the below macros look convoluted, their purpose is simple: wrap the 
% 'b@<citation-key>' macro in another one called 'scholar@b@<citation-key>', 
% which just calls a coloring macro followed by 'b@<citation-key>'.
% \def\scholar@b@csname{b@#1}
% \def\scholar@inner@b@csname{scholar@inner@b@#1}
% \expandafter\let%
%   \csname \scholar@inner@b@csname\expandafter\endcsname%
%   \csname \scholar@b@csname\endcsname%
% \expandafter\edef%
%   \csname \scholar@b@csname\endcsname %
%   {\noexpand\textcolor[rgb]{#2,#3,#4}{\csname \scholar@inner@b@csname\endcsname}}
```
